Use ConfigurationManager in Apple provider

martincostello opened the pull request:
Switch to using the `ConfigurationManager` and Apple's OpenID Connect discovery endpoint instead of the custom implementation.

Also changes `PrivateKeyBytes` to accept a `CancellationToken`.

Resolves #421.

martincostello committed:
Use ConfigurationManager in Apple provider

Switch to using the ConfigurationManager and Apple's OpenID Connect discovery endpoint instead of the custom implementation.
Also changes PrivateKeyBytes to accept a CancellationToken.
Resolves #421.

kevinchalet commented on src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleIdTokenValidator.cs:
```diff
@@ -5,62 +5,67 @@
+                throw new InvalidOperationException($"Token validation parameters have not been set on the {nameof(AppleAuthenticationOptions)} instance.");
+            }
+
+            OpenIdConnectConfiguration configuration = await context.Options.ConfigurationManager.GetConfigurationAsync(context.HttpContext.RequestAborted);
```
You need moar `var` love in your life 😃

kevinchalet commented on src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleIdTokenValidator.cs:
```diff
@@ -5,62 +5,67 @@
+
+            OpenIdConnectConfiguration configuration = await context.Options.ConfigurationManager.GetConfigurationAsync(context.HttpContext.RequestAborted);
+
+            context.Options.TokenValidationParameters.IssuerSigningKeys = configuration.JsonWebKeySet.Keys;
```
Assuming this method can be concurrently called, this is a risky operation. Instead, you may want to clone `context.Options.TokenValidationParameters` and mutate the cloned instance.

kevinchalet commented on src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleIdTokenValidator.cs:
```diff
@@ -5,62 +5,67 @@
-                context.Options.JwtSecurityTokenHandler.ValidateToken(context.IdToken, parameters, out _);
+                var result = context.Options.SecurityTokenHandler.ValidateToken(context.IdToken, context.Options.TokenValidationParameters);
+
+                if (result.Exception != null || !result.IsValid)
```
`!=` -> `is not`? 😃

martincostello committed:
Clone the validation parameters

Clone the token validation parameters, rather than mutating it.

martincostello committed:
Fix client secret generation

Seems that compared to JsonWebTokenHandler, JwtSecurityTokenHandler did something under the hood that made "SecurityAlgorithms.EcdsaSha256Signature" be encoded as "SecurityAlgorithms.EcdsaSha256" instead.
This was then causing Apple to reject the requests to the token endpoint.
To fix that, explicitly use the correct algorithm name.

martincostello committed:
Remove unused package reference

Remove unused package reference for System.IdentityModel.Tokens.Jwt.

martincostello commented on src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleClientSecretGenerator.cs:
```diff
@@ -133,7 +130,7 @@ private SigningCredentials CreateSigningCredentials(string keyId, ECDsa algorith
             // Use a custom CryptoProviderFactory so that keys are not cached and then disposed of, see below:
             // https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1302
-            return new SigningCredentials(key, SecurityAlgorithms.EcdsaSha256Signature)
+            return new SigningCredentials(key, SecurityAlgorithms.EcdsaSha256)
```
This change is because when `JwtSecurityTokenHandler` was used, its internals include some code to coerce algorithm names to be those supported for JWTs:

  * [Mapping table](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs#L82)
  * [Usage](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs#L551)
  * [Coercion](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/System.IdentityModel.Tokens.Jwt/JwtHeader.cs#L101-L102)

`JsonWebTokenHandler` does not do that, and [uses the algorithm name verbatim](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L147), which was causing the wrong algorithm name to be sent to Apple, causing it to be rejected.

martincostello commented:
Are you happy with this as-is Kévin?

kevinchalet commented:
Looks great! 👍🏻
